recipes-devtools/cargo/cargo-cross-canadian_1.49.0.bb

```diff
@@ -0,0 +1,6 @@
+require recipes-devtools/rust/rust-source-${PV}.inc
+require recipes-devtools/rust/rust-snapshot-${PV}.inc
+
+FILESEXTRAPATHS_prepend := "${THISDIR}/cargo-${PV}:"
+
+require cargo-cross-canadian.inc
```

Comment:
Thanks for joining on this work! I'd wonder if there's more people to get this feature as the issue has been filed a couple of years ago already.

To the actual issue - could you please share which branch you base on and which layers are included. The description should probably mention that it's currently about dunfell and upwards.  If you've already some fix in mind, please PR against the feature branch or just send a patch. Otherwise please let me know how it can be reproduced, maybe that's the base poky branch or something in the config that i currently don't have.

Regrading the question about the canadian cross build - the produced binaries are dedicated to the prospective host architecture.  It is not agnostic. As meta-rust only supports amd64 as a build host, it might make things simpler, but still keeping up with teh concept is good. For the same reason, no crosssdk builds are necessary.

Thanks

Reply:
> Thanks for joining on this work! I'd wonder if there's more people to get this feature as the issue has been filed a couple of years ago already.

I was also surprised to see that someone else started working on this simultaneously (in particular because I checked the open PRs before starting this off).

> To the actual issue with the header - could you please share which branch you base on and which layers are included. The description should probably mention that it's currently about dunfell and upwards. If you've already some fix in mind, please PR against the feature branch or just send a patch. Otherwise please let me know how it can be reproduced, maybe that's the base poky branch or something in the config that i currently don't have.

I'm testing this on dunfell (although I'd probably end up using this on earlier versions as well). At the end of the day I suspect that this boils down to the missing dependency on `nativesdk-libssh2` (although in that case I'm also having trouble understanding why it compiles for everybody else). I don't have a build dir ready to retrieve logs but I'll revisit this later.

> Regrading the question about the canadian cross build - the produced binaries are dedicated to the prospective host architecture. It is not agnostic. As meta-rust only supports amd64 as a build host, it might make things simpler, but still keeping up with teh concept is good. For the same reason, no crosssdk builds are necessary.

I'm having a hard to time to find the correct Yocto terminology for this as (in contrast to GCC) a rustc compiled for `SDK_SYS` seems to be perfectly capable to compile for any other architecture which is supported by its LLVM and for which a `std` build is available (and I'd expect cargo to be even more flexible here).

When starting off on this quest I even attempted to configure `rust-native` in such a way that it would produce binaries for `BUILD_SYS` / `SDK_SYS` and the target's architecture (however these attempts failed due to various assumptions in the Yocto toolchain which prevented me from installing the `cross` and `crosssdk` GCC variants).

So far I ended up packaging the compiler parts in a `nativesdk-rust` package and the `std` parts in a `nativesdk-libstd-rs` package. In this setup the `cross-canadian-rust` package only provides a target configuration file (similar to the approach chosen for the current `cross` recipe) and a linker wrapper script (as at least in my setup the linker is not able to find its support libraries without explicitly passing it via `--sysroot`).

Reply:
Yes, the Rust builds stage2 are in general capable of cross compiling. What they need is:

- `rustc`
- libstd for host
- libstd for target

If desired, a custom target json can be used, too.

In case host and target are same, obviously one can get away just with one version of `libstd`.

Thanks.

Reply:
So I've managed to fix the compilation issue by manually adding `nativesdk-libssh2` to `DEPENDS`. I am doing all my Yocto builds in a minimal Debian 10 container so I am wondering if you maybe didn't see this because of host contamination.

After that I was hit with a packaging QA warning complaining about a missing provider for libcurl

`ERROR: cargo-cross-canadian-aarch64-1.49.0-r0 do_package_qa: QA Issue: /opt/sdk/sysroots/x86_64-pokysdk-linux/usr/bin/cargo contained in package cargo-cross-canadian-aarch64 requires libcurl.so.4()(64bit), but no providers found in RDEPENDS_cargo-cross-canadian-aarch64? [file-rdeps]`

Adding `nativesdk-libcurl` to `DEPENDS` also solved this (however I would've expected the build to fail because of the missing dependency so it might make sense to double check which headers the build is actually using).